src/libraries/Common/src/Interop/OSX/Swift.Runtime/CryptoKit.cs

```diff
@@ -0,0 +1,274 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.Swift;
+
+#pragma warning disable CS3016 // Arrays as attribute arguments are not CLS Compliant
+#pragma warning disable SYSLIB1051
+
+namespace Swift.Runtime
+{
+    /// <summary>
+    /// Represents ChaChaPoly in C#.
+    /// </summary>
+    internal unsafe partial struct ChaChaPoly
+    {
+        /// <summary>
+        /// Represents Nonce in C#.
+        /// </summary>
+        internal sealed unsafe partial class Nonce
+        {
+            private const int _payloadSize = 16;
+            internal void* payload;
+
+            internal Nonce()
+            {
+                payload = Marshal.AllocHGlobal(_payloadSize).ToPointer();
+                SwiftIndirectResult swiftIndirectResult = new SwiftIndirectResult(payload);
+                CryptoKit.PInvoke_ChaChaPoly_Nonce_Init(swiftIndirectResult);
+            }
+
+            internal Nonce(Data data)
+            {
+                payload = Marshal.AllocHGlobal(_payloadSize).ToPointer();
+                SwiftIndirectResult swiftIndirectResult = new SwiftIndirectResult(payload);
+                void* dataPtr = &data;
+
+                CryptoKit.PInvoke_ChaChaPoly_Nonce_Init2(swiftIndirectResult, dataPtr, data.Metadata(), default(DataProtocol).WitnessTable(data), out SwiftError error);
+
+                if (error.Value != null) // Should throw when nonce is not 12 bytes
+                {
+                    throw new Exception();
```

Comment:
Would it possible to get more descriptive exceptions? 

For example, taking the Swift error and based on its type propagating the exception to C#? Based on my knowledge that would probably require another call to Swift to check the error type. However, maybe it is possible to retrieve the metadata directly using the `error.Value`?

Reply:
Good idea. I'm not sure if there's a more specific exception we can throw on the C# side than `CryptographicException`.

We plan to project the `CryptoKitError` enum, and then we will be able to completely eliminate the Swift wrapper. I suggest avoiding the projection of the `CryptoKitError` for now to keep the scope of this change limited.